### PR TITLE
Implement new redshift outputter

### PR DESCRIPTION
**Pull request by ChrisPRobinson**
1) Simplified Redshift DateTable and ID augmenter
2) Added has_keys to table schema to make it easier to see which columns
exist
3) Fixed bugs in transform date table id transformer
4) Added a Test incrementing id generator
5) Added new Redshift outputter
6) Added tests for Redshift2 outputter and fixed tests.

### CODE REVIEW

File: lib/etl/redshift/id_augmenter.rb [hunk 1]
```diff
@@ -2,14 +2,13 @@
 
 module ETL::Redshift
   class IDAugmenter <::ETL::Transform::IDAugmenter
-    def initialize(client, table_name, natural_keys, filter_part = nil, id_generator = nil)
-      table_schema = client.table_schema(table_name)
+    def initialize(client, table_schema, natural_keys, filter_part, id_generator)
       pks = table_schema.primary_key
       raise "Support converting multiple pks to one dw pk on #{table_name}" if pks.count > 1
```
Comment (emya) [inline lib/etl/redshift/id_augmenter.rb]: `table_name `

File: lib/etl/transform/id_augmenter.rb [hunk 2]
```diff
@@ -3,14 +3,25 @@
...
 module ETL::Transform
 
+  # used for testing only
+  class IncrementingTestIDGenerator
+    def initialize(start=0)
+      @count = start
+    end
+    def generate_id
+      @count = @count  + 1
+      @count
```
Comment (emya) [inline lib/etl/transform/id_augmenter.rb]: just `@count += 1`

Comment (ChrisPRobinson) [reply lib/etl/transform/id_augmenter.rb]: fixed

File: lib/etl/transform/id_augmenter.rb [hunk 3]
```diff
@@ -3,14 +3,24 @@
 
 module ETL::Transform
 
+  # used for testing only
+  class IncrementingTestIDGenerator
```
Comment (emya) [inline lib/etl/transform/id_augmenter.rb]: I think this should be in spec, not this file?

**Comment by ChrisPRobinson** [comment]
Back to wip. have an implementation that should work after I get things back to tested.

File: lib/etl/output/redshift2.rb [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+      @main_table_name = main_table_name
+      @history_table_name = optional_history_table_name
+      @scd_columns = scd_columns
+      @id_generator = id_generator || IDUUIDGenerator.new
+      @now_generator = ::ETL::Output::CurrentDateTimeGenerator.new
+    end
+
+    # Runs the ETL job
+    def run_internal
+      raise "main_table_name must not be empty" if @main_table_name.empty?
```
Comment (emya) [inline lib/etl/output/redshift2.rb]: This should be done in `initialize` like `DataHistoryRowTransformer`?

File: lib/etl/output/redshift2.rb [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+      raise "data table #{@main_table_schema.name} must only have one pk" if !@main_table_schema.primary_key.count == 1
+      raise "history table #{@history_table_schema.name} must only have one pk" if !@history_table_schema.primary_key.count == 1
+      main_table_surrogate_key = @main_table_schema.primary_key[0]
+      history_table_key = @history_table_schema.primary_key[0]
+      # If no surrogate key exists then none was found so its new.
+      if !named_rows[@main_table_schema.name].has_key?(main_table_surrogate_key)
+        # This is a new row so need the following
+        #    a. new row for orgs
+        #    b. new row for orgs_history
```
Comment (emya) [inline lib/etl/output/redshift2.rb]: Tiny thing, but comments are very specific to `orgs` jobs that we have in DW repo.

File: lib/etl/output/redshift2.rb [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+            break
+          end
+        end
+
+        if scd_changed
+          # When a slowly changing dimension hash changed the following needs to occur
+            #    a. update current history row to not be current and add end_date
+            #    b. add new row for history with the start time to now and it being current.
+            existing_history_row = named_rows[@history_table_schema.name]
+            puts "existing: #{existing_history_row.inspect}"
```
Comment (emya) [inline lib/etl/output/redshift2.rb]: Remove 'puts'

**Pull request merged**